tianshou/data/buffer/her.py

```diff
@@ -0,0 +1,180 @@
+
+    def set_batch(self, batch: Batch) -> None:
+        self._restore_cache()
+        return super().set_batch(batch)
+
+    def update(self, buffer: Union["HERReplayBuffer", "ReplayBuffer"]) -> np.ndarray:
+        self._restore_cache()
+        return super().update(buffer)
+
+    def add(
+        self,
+        batch: Batch,
+        buffer_ids: Optional[Union[np.ndarray, List[int]]] = None
+    ) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
+        self._restore_cache()
+        return super().add(batch, buffer_ids)
+
+    def sample_indices(self, batch_size: int) -> np.ndarray:
+        """Get a random sample of index with size = batch_size.
+
+        Return all available indices in the buffer if batch_size is 0; return an \
+        empty numpy array if batch_size < 0 or no available index can be sampled. \
+        Additionally, some episodes of the sampled transitions will be re-written \
+        according to HER.
+        """
+        self._restore_cache()
+        indices = np.sort(super().sample_indices(batch_size=batch_size))
+        self.rewrite_transitions(indices)
+        return indices
+
+    def rewrite_transitions(self, indices: np.ndarray) -> None:
+        """Re-write the goal of some sampled transitions' episodes according to HER.
+
+        Currently applies only HER's 'future' strategy. The new goals will be written \
+        directly to the internal batch data temporarily and will be restored right \
+        before the next sampling or when using some of the buffer's method (e.g. \
+        `add`, `save_hdf5`, etc.). This is to make sure that n-step returns \
+        calculation etc., performs correctly without additional alteration.
+        """
+        if indices.size == 0:
+            return
+        # Construct episode trajectories
+        indices = [indices]
+        for _ in range(self.horizon - 1):
+            indices.append(self.next(indices[-1]))
+        indices = np.stack(indices)
+
+        # Calculate future timestep to use
+        current = indices[0]
+        terminal = indices[-1]
+        future_offset = np.random.uniform(size=len(indices[0])) * (terminal - current)
+        future_offset = future_offset.astype(int)
+        future_t = (current + future_offset)
+
+        # Compute indices
+        #   open indices are used to find longest, unique trajectories among
+        #   presented episodes
+        unique_ep_open_indices = np.unique(terminal, return_index=True)[1]
+        unique_ep_indices = indices[:, unique_ep_open_indices]
+        #   close indices are used to find max future_t among presented episodes
+        unique_ep_close_indices = np.hstack(
+            [(unique_ep_open_indices - 1)[1:],
+             len(terminal) - 1]
+        )
+        #   episode indices that will be altered
+        her_ep_indices = np.random.choice(
+            len(unique_ep_open_indices),
+            size=int(len(unique_ep_open_indices) * self.future_p),
```

Comment:
Could you please briefly explain why we need `self.future_p` here, instead of all indices in this episode?

Reply:
Yes, you are correct. HER only performs goal re-writing on some random episodes (the amount is depending on the ratio parameter `future_k`, or in here, probablity `future_p`).